asapdiscovery-docking/asapdiscovery/docking/schema/pose_generation.py

```diff
@@ -0,0 +1,198 @@
+    type: Literal["OpenEyeConstrainedPoseGenerator"] = "OpenEyeConstrainedPoseGenerator"
+    max_confs: PositiveInt = Field(
+        1000, description="The maximum number of conformers to try and generate."
+    )
+    energy_window: PositiveFloat = Field(
+        20,
+        description="Sets the maximum allowable energy difference between the lowest and the highest energy conformers, in units of kcal/mol",
+    )
+    clash_cutoff: PositiveFloat = Field(2.0, description="The ")
+
+    def provenance(self) -> dict[str, Any]:
+        from openeye import oechem, oeomega
+
+        return {
+            "oechem": oechem.OEChemGetVersion(),
+            "oeomega": oeomega.OEOmegaGetVersion(),
+        }
+
+    def _generate_poses(
+        self,
+        receptor: PreppedTarget,
+        reference_ligand: ReferenceLigand,
+        ligands: list[Ligand],
+        core_smarts: str,
+    ) -> list[Ligand]:
+        """Use openeye oeomega to generate constrained poses for ligands"""
+
+        from openeye import oechem, oeomega
+
+        # Make oechem be quiet
+        oechem.OEThrow.SetLevel(oechem.OEErrorLevel_Quiet)
+
+        ref_mol = reference_ligand.to_oemol()
+        # Prep the substructure searching
+        ss = oechem.OESubSearch(core_smarts)
+        oechem.OEPrepareSearch(ref_mol, ss)
+        core_fragment = None
+
+        for match in ss.Match(ref_mol):
+            core_fragment = oechem.OEGraphMol()
+            oechem.OESubsetMol(core_fragment, match)
+            break
+
+        if core_fragment is None:
+            raise RuntimeError(
+                f"A core fragment could not be extracted from the reference ligand using core smarts {core_smarts}"
+            )
+
+        # Create an Omega instance
+        omega_opts = oeomega.OEOmegaOptions(oeomega.OEOmegaSampYEling_Dense)
+        # Set the fixed reference molecule
+        omega_fix_opts = oeomega.OEConfFixOptions()
+        omega_fix_opts.SetFixMaxMatch(10)  # allow multiple MCSS matches
+        omega_fix_opts.SetFixDeleteH(True)  # only use heavy atoms
+        omega_fix_opts.SetFixMol(core_fragment)
+        omega_fix_opts.SetFixRMS(1.0)
+        # set the matching atom and bond expressions
+        atomexpr = oechem.OEExprOpts_Aromaticity | oechem.OEExprOpts_AtomicNumber
+        bondexpr = oechem.OEExprOpts_BondOrder | oechem.OEExprOpts_Aromaticity
+        omega_fix_opts.SetAtomExpr(atomexpr)
+        omega_fix_opts.SetBondExpr(bondexpr)
+        omega_opts.SetConfFixOptions(omega_fix_opts)
+        # set the builder options
+        mol_builder_opts = oeomega.OEMolBuilderOptions()
+        mol_builder_opts.SetStrictAtomTypes(
+            False
+        )  # don't give up if MMFF types are not found
+        omega_opts.SetMolBuilderOptions(mol_builder_opts)
+        omega_opts.SetWarts(False)  # expand molecule title
+        omega_opts.SetStrictStereo(True)  # set strict stereochemistry
```

Comment:
This will cause undefined to fail, may want a nicer warning prior. Could be a check on `Ligand`

Reply:
Yeah I wasn't sure if this component should do that or if the workflow we build would do that in between stages something like take in molecules deduplicate inputs -> state enumeration if requested -> check ligands -> all other stages?